### PR TITLE
udev: Add rules for accelerometer orientation quirks

### PR DESCRIPTION
**Pull request by hadess**
This commit adds a rules file to extract the properties from hwdb
to set on i2c IIO devices. This is used to set the ACCEL_MOUNT_MATRIX
property on IIO devices, to be consumed by iio-sensor-proxy or
equivalent daemon.

The hwdb file contains documentation on how to write quirks. Note
however that mount information is usually exported in:
- the device-tree for ARM devices
- the ACPI DSDT for Intel-compatible devices
but currently not extracted by the kernel.

Also note that some devices have the framebuffer rotation that changes
between the bootloader and the main system, which might mean that the
accelerometer is then wrongly oriented. This is a missing feature in the
i915 kernel driver: https://bugs.freedesktop.org/show_bug.cgi?id=94894
which needs to be fixed, and won't require quirks.

### CODE REVIEW

**Comment by hadess** [comment]
Here for review. I've had my head buried in this for a little while, so the documentation might be nonsensical.

File: rules/60-sensor.rules [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="sensor_end"
+KERNEL!="i2c*", GOTO="sensor_end"
```
Comment (hadess) [inline rules/60-sensor.rules]: I would like to match on iio devices instead, but they don't have a modalias themselves. If there was a way to match the parent device instead.

This is what an IIO device's tree might look like when not connected via I2C:
```
P: /devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/0003:273F:1008.001A
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/0003:273F:1008.001A
E: DRIVER=hid-sensor-hub
E: HID_ID=0003:0000273F:00001008
E: HID_NAME=Hughski Ltd. ColorHugALS
E: HID_PHYS=usb-0000:00:14.0-1/input0
E: HID_UNIQ=
E: MODALIAS=hid:b0003g0003v0000273Fp00001008
E: SUBSYSTEM=hid

P: /devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/0003:273F:1008.001A/HID-SENSOR-200041.2.auto
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/0003:273F:1008.001A/HID-SENSOR-200041.2.auto
E: DEVTYPE=mfd_device
E: DRIVER=hid_sensor_als
E: MODALIAS=platform:HID-SENSOR-200041
E: SUBSYSTEM=platform

P: /devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/0003:273F:1008.001A/HID-SENSOR-200041.2.auto/iio:device0
N: iio:device0
E: DEVNAME=/dev/iio:device0
E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/0003:273F:1008.001A/HID-SENSOR-200041.2.auto/iio:device0
E: DEVTYPE=iio_device
E: IIO_SENSOR_PROXY_TYPE=iio-buffer-als
E: MAJOR=240
E: MINOR=0
E: SUBSYSTEM=iio
E: SYSTEMD_WANTS=iio-sensor-proxy.service
E: TAGS=:systemd:
E: USEC_INITIALIZED=98015640898
```

**Comment by hadess** [comment]
```
-KERNELS=="i2c*", IMPORT{builtin}="hwdb 'sensor:modalias:$attr{modalias}:$attr{[dmi/id]modalias}'", \
+KERNELS=="iio*", IMPORT{builtin}="hwdb 'sensor:modalias:$attr{[../]modalias}:$attr{[dmi/id]modalias}'", \
```
?? I have no idea.

**Comment by poettering** [comment]
normally it should suffice to use SUBSYSTEMS= (in plural) to match a parent of the current device, and the attribute checks are than towards that parent device automatically.

**Comment by hadess** [comment]
> normally it should suffice to use SUBSYSTEMS= (in plural) to match a parent of the current device, and the attribute checks are than towards that parent device automatically.

For accelerometers, the main device is IIO (or input, but we'll get to that) and the parent or grand-parent is the i2c, or USB, or whatever else device.

I'd like to match for IIO devices, but do the attributes check against the parent, as the parent has a modalias, because it's a physical device. In English, this would be:
- device's subsystem is IIO
- read the parent's (or grand-parent's) modalias and the system dmi
- look for hwdb matches for this combination
- assign the result to the IIO device

Maybe it would be easier to assign the quirk to the device itself, but that would mean having information about the accelerometer checked for every usb device, and attached to the physical device. I'd rather it was attached to the actual iio device that's an accelerometer...

**Comment by poettering** [comment]
```
SUBSYSTEM=="iio", SUBSYSTEMS=="usb|pci|i2c", …
```

should do what you want, no? It matches on devices of the subsystem IO, then goes up the tree to find a usb or pci or i2c device, and then proceeds from there... If you the add further rule components to the same line they test the parent device...

**Comment by hadess** [comment]
I'll try those out, thanks.

**Comment by hadess** [comment]
That works! I fixed a type in the example string in the hwdb file, and updated the rules file, which should now work on all the IIO devices. I'll expect @phomes to update it for accelerometer input devices soon.

**Comment by phomes** [comment]
For my input device I guess I have to do things slightly differently since I don't have a modalias? This is what I have to match:
udevadm info -a /dev/input/event7
...
  looking at device '/devices/platform/lis3lv02d/input/input13/event7':
    KERNEL=="event7"
    SUBSYSTEM=="input"
    DRIVER==""

  looking at parent device '/devices/platform/lis3lv02d/input/input13':
    KERNELS=="input13"
    SUBSYSTEMS=="input"
    DRIVERS==""
    ATTRS{max}=="2000"
    ATTRS{min}=="0"
    ATTRS{name}=="ST LIS3LV02DL Accelerometer"
    ATTRS{phys}=="lis3lv02d/input0"
    ATTRS{poll}=="50"
    ATTRS{properties}=="0"
    ATTRS{uniq}==""

  looking at parent device '/devices/platform/lis3lv02d':
    KERNELS=="lis3lv02d"
    SUBSYSTEMS=="platform"
    DRIVERS==""
    ATTRS{driver_override}=="(null)"
    ATTRS{position}=="(72,198,1044)"
    ATTRS{rate}=="100"
    ATTRS{selftest}=="OK -30 29 34"

  looking at parent device '/devices/platform':
    KERNELS=="platform"
    SUBSYSTEMS==""
    DRIVERS==""

One way would be to go for the kernels but that gets quite specific for this device:
rules:
SUBSYSTEM=="input", SUBSYSTEMS=="platform|input", KERNELS=="lis3lv02d", \
  IMPORT{builtin}="hwdb 'sensor:modalias:$attr{modalias}:$attr{[dmi/id]modalias}'", \
  GOTO="sensor_end"

hwdb:
sensor:modalias:*:dmi:*svn*Hewlett-Packard*:*pn*HPEliteBook8540w*
  ACCEL_MOUNT_MATRIX=1, 0, 0; 0, 0, -1; 0, 1, 0

From udev-builtin-input_id we already have ID_INPUT_ACCELEROMETER set for the device. It might be more general to do:
SUBSYSTEM=="input", SUBSYSTEMS=="platform|input", ENV{ID_INPUT_ACCELEROMETER}=="1", \
  IMPORT{builtin}="hwdb 'sensor:modalias:$attr{modalias}:$attr{[dmi/id]modalias}'", \
  GOTO="sensor_end"

**Comment by hadess** [comment]
You do have a modalias, you listed it in https://github.com/hadess/iio-sensor-proxy/pull/111#issuecomment-264547522 :
```
E: MODALIAS=input:b0019v0000p0000e0000-e0,3,kra0,1,2,mlsfw
```

**Comment by hadess** [comment]
I added the Makefile rules now.

**Comment by hadess** [comment]
I couldn't figure out how to add support for the new hwdb properties and matches to `hwdb/parse_hwdb.py`. If anyone has ideas...

**Comment by phomes** [comment]
The parent has the modalias. Is there a way to match on that?

**Comment by hadess** [comment]
> The parent has the modalias. Is there a way to match on that?

That's already what the iio match rule does when we run the "SUBSYSTEMS" check. Ever attribute we get after that is for an ancestor with one of those subsystems.

**Comment by phomes** [comment]
I guess I am doing something dumb then. With the modalias set to anything but just a * I cannot get it to match on the input/event. Only on the parent. With modalias as * they both get it. 

Here are my current rules and hwdb entries:

SUBSYSTEM=="input", SUBSYSTEMS=="platform|input", \
  IMPORT{builtin}="hwdb 'sensor:modalias:$attr{modalias}:$attr{[dmi/id]modalias}'", \
  GOTO="sensor_end"

sensor:modalias:input:b0019v0000p0000*:dmi:*svn*Hewlett-Packard*:*pn*HPEliteBook8540w*
  ACCEL_MOUNT_MATRIX=1, 0, 0; 0, 0, -1; 0, 1, 0

udevadm info /dev/input/event7
P: /devices/platform/lis3lv02d/input/input13/event7
N: input/event7
S: input/by-path/platform-lis3lv02d-event
E: DEVLINKS=/dev/input/by-path/platform-lis3lv02d-event
E: DEVNAME=/dev/input/event7
E: DEVPATH=/devices/platform/lis3lv02d/input/input13/event7
E: ID_INPUT=1
E: ID_INPUT_ACCELEROMETER=1
E: ID_PATH=platform-lis3lv02d
E: ID_PATH_TAG=platform-lis3lv02d
E: IIO_SENSOR_PROXY_TYPE=input-accel
E: LIBINPUT_DEVICE_GROUP=19/0/0/0:lis3lv02d
E: MAJOR=13
E: MINOR=71
E: SUBSYSTEM=input
E: SYSTEMD_WANTS=iio-sensor-proxy.service
E: TAGS=:systemd:
E: USEC_INITIALIZED=11606618

udevadm info --path=/devices/platform/lis3lv02d/input/input13
P: /devices/platform/lis3lv02d/input/input13
E: ABS=7
E: ACCEL_MOUNT_MATRIX=1, 0, 0; 0, 0, -1; 0, 1, 0
E: DEVPATH=/devices/platform/lis3lv02d/input/input13
E: EV=9
E: ID_FOR_SEAT=input-platform-lis3lv02d
E: ID_INPUT=1
E: ID_INPUT_ACCELEROMETER=1
E: ID_PATH=platform-lis3lv02d
E: ID_PATH_TAG=platform-lis3lv02d
E: IIO_SENSOR_PROXY_TYPE=input-accel
E: MODALIAS=input:b0019v0000p0000e0000-e0,3,kra0,1,2,mlsfw
E: NAME="ST LIS3LV02DL Accelerometer"
E: PHYS="lis3lv02d/input0"
E: PRODUCT=19/0/0/0
E: PROP=0
E: SUBSYSTEM=input
E: SYSTEMD_WANTS=iio-sensor-proxy.service
E: TAGS=:seat:systemd:

**Comment by keszybz** [comment]
Force-pushed with two changes:
- use just one space for indentation,
- update hwdb_parser to support the new property.

**Comment by hadess** [comment]
Thanks @keszybz. I guess we can look at supporting @phomes ' system in a separate PR.

**Comment by hadess** [comment]
Moved the discussion about phomes' hardware back to https://github.com/hadess/iio-sensor-proxy/pull/111

**Pull request merged**